File: impl-large/python/store.py
"""Task Service, large tier — the in-memory state and its repositories."""

from domain import (
    DEFAULT_QUOTA, PROBE_QUOTA, AuditEntry, Comment, OutboxEvent, Project, Session, Task, User,
)

users: dict[int, User] = {}
sessions: dict[str, Session] = {}
projects: dict[int, Project] = {}
tasks: dict[int, Task] = {}
comments: dict[int, Comment] = {}
audit: list[AuditEntry] = []
outbox: list[OutboxEvent] = []
idempotency: dict[tuple[str, str], tuple[int, dict]] = {}
by_status: dict[int, int] = {}
by_route: dict[str, int] = {}

requests = 0
next_project_id = 1
next_task_id = 1
next_comment_id = 1
next_user_id = 5
next_seq = 1


def seed() -> None:
    users.update({
        1: User(1, "admin", "admin-secret", "admin", DEFAULT_QUOTA),
        2: User(2, "alice", "alice-secret", "user", DEFAULT_QUOTA),
        3: User(3, "bob", "bob-secret", "user", DEFAULT_QUOTA),
        4: User(4, "probe", "probe-secret", "user", PROBE_QUOTA),
    })


def take_seq() -> int:
    global next_seq
    value = next_seq
    next_seq += 1
    return value


def record(actor_id: int, action: str, resource: str, resource_id: int) -> None:
    """Append one audit entry and one outbox event for a successful write."""
    audit.append(AuditEntry(take_seq(), actor_id, action, resource, resource_id))
    outbox.append(OutboxEvent(take_seq(), f"{resource}.{action}", resource_id))


def count_request(route: str, status: int) -> None:
    global requests
    requests += 1
    by_route[route] = by_route.get(route, 0) + 1
    by_status[status] = by_status.get(status, 0) + 1


def find_user(user_id: int, include_deleted: bool = False) -> User | None:
    user = users.get(user_id)
    if user is None or (user.deleted and not include_deleted):
        return None
    return user


def find_by_username(username: str) -> User | None:
    return next((u for u in users.values() if u.username == username and not u.deleted), None)


def insert_user(username: str, password: str, role: str, quota: int) -> User:
    global next_user_id
    user = User(next_user_id, username, password, role, quota)
    users[user.id] = user
    next_user_id += 1
    return user


def find_project(project_id: int, include_deleted: bool = False) -> Project | None:
    project = projects.get(project_id)
    if project is None or (project.deleted and not include_deleted):
        return None
    return project


def insert_project(name: str, owner_id: int) -> Project:
    global next_project_id
    project = Project(next_project_id, name, owner_id)
    projects[project.id] = project
    next_project_id += 1
    return project


def find_task(task_id: int, include_deleted: bool = False) -> Task | None:
    task = tasks.get(task_id)
    if task is None or (task.deleted and not include_deleted):
        return None
    return task


def insert_task(project_id: int, title: str, priority: int, assignee_id: int | None,
                internal_note: str) -> Task:
    global next_task_id
    task = Task(next_task_id, project_id, title, priority, "todo", assignee_id, internal_note)
    tasks[task.id] = task
    next_task_id += 1
    return task


def find_comment(comment_id: int) -> Comment | None:
    return comments.get(comment_id)


def insert_comment(task_id: int, author_id: int, body: str) -> Comment:
    global next_comment_id
    comment = Comment(next_comment_id, task_id, author_id, body)
    comments[comment.id] = comment
    next_comment_id += 1
    return comment


def live_tasks_of(project_id: int) -> list[Task]:
    return [t for t in tasks.values() if t.projectId == project_id and not t.deleted]


def task_count(project_id: int) -> int:
    return len(live_tasks_of(project_id))


def outbox_pending() -> int:
    return sum(1 for event in outbox if not event.delivered)

File: impl-large/ruby/store.rb
# frozen_string_literal: true

# Task Service, large tier — the in-memory state and its repositories.

require_relative "domain"

# `extend self` turns every method below into a module method, so the state can
# sit in the module's own instance variables and still be reached as Store.users
# from the layers above.
module Store
  extend self

  attr_reader :users, :sessions, :projects, :tasks, :comments, :audit, :outbox,
              :idempotency, :by_status, :by_route, :requests

  @users = {}
  @sessions = {}
  @projects = {}
  @tasks = {}
  @comments = {}
  @audit = []
  @outbox = []
  @idempotency = {}
  @by_status = {}
  @by_route = {}

  @requests = 0
  @next_project_id = 1
  @next_task_id = 1
  @next_comment_id = 1
  @next_user_id = 5
  @next_seq = 1

  def seed
    @users.update(
      1 => Domain::User.new(1, "admin", "admin-secret", "admin", Domain::DEFAULT_QUOTA, 1, false),
      2 => Domain::User.new(2, "alice", "alice-secret", "user", Domain::DEFAULT_QUOTA, 1, false),
      3 => Domain::User.new(3, "bob", "bob-secret", "user", Domain::DEFAULT_QUOTA, 1, false),
      4 => Domain::User.new(4, "probe", "probe-secret", "user", Domain::PROBE_QUOTA, 1, false)
    )
  end

  def take_seq
    value = @next_seq
    @next_seq += 1
    value
  end

  # Append one audit entry and one outbox event for a successful write.
  def record(actor_id, action, resource, resource_id)
    @audit << Domain::AuditEntry.new(take_seq, actor_id, action, resource, resource_id)
    @outbox << Domain::OutboxEvent.new(take_seq, "#{resource}.#{action}", resource_id, false)
  end

  def count_request(route, status)
    @requests += 1
    @by_route[route] = @by_route.fetch(route, 0) + 1
    @by_status[status] = @by_status.fetch(status, 0) + 1
  end

  def find_user(user_id, include_deleted = false)
    user = @users[user_id]
    user if user && (include_deleted || !user.deleted)
  end

  def find_by_username(username)
    @users.values.find { |user| user.username == username && !user.deleted }
  end

  def insert_user(username, password, role, quota)
    user = Domain::User.new(@next_user_id, username, password, role, quota, 1, false)
    @users[user.id] = user
    @next_user_id += 1
    user
  end

  def find_project(project_id, include_deleted = false)
    project = @projects[project_id]
    project if project && (include_deleted || !project.deleted)
  end

  def insert_project(name, owner_id)
    project = Domain::Project.new(@next_project_id, name, owner_id, 1, false)
    @projects[project.id] = project
    @next_project_id += 1
    project
  end

  def find_task(task_id, include_deleted = false)
    task = @tasks[task_id]
    task if task && (include_deleted || !task.deleted)
  end

  def insert_task(project_id, title, priority, assignee_id, internal_note)
    task = Domain::Task.new(@next_task_id, project_id, title, priority, "todo",
                            assignee_id, internal_note, 1, false)
    @tasks[task.id] = task
    @next_task_id += 1
    task
  end

  def find_comment(comment_id)
    @comments[comment_id]
  end

  def insert_comment(task_id, author_id, body)
    comment = Domain::Comment.new(@next_comment_id, task_id, author_id, body)
    @comments[comment.id] = comment
    @next_comment_id += 1
    comment
  end

  def live_tasks_of(project_id)
    @tasks.values.select { |task| task.project_id == project_id && !task.deleted }
  end

  def task_count(project_id)
    live_tasks_of(project_id).length
  end

  def outbox_pending
    @outbox.count { |event| !event.delivered }
  end
end
